lib/experimental/text_layouter.nim

```diff
@@ -0,0 +1,1754 @@
+## Nim reimplementation of layout algorithm from
+## https://github.com/google/rfmt.
+
+# Thanks to nim being syntactically close to python this is mostly
+# just blatant copy-paste of python code with added type annotations.
+
+import std/[
+  strutils,
+  sequtils,
+  macros,
+  tables,
+  strformat,
+  lenientops,
+  options,
+  hashes,
+  math,
+  sugar,
+  intsets
+]
+
+const
+  infty = 1024 * 1024 * 1024 * 1024
+
+func inf(a: int): bool = (infty - 4096 <= a) and (a <= infty + 4096)
+
+func `*`(a: SomeNumber, b: bool): SomeNumber = (if b: a else: 0)
+func get[T](inseq: seq[Option[T]]): seq[T] =
+  for elem in inseq:
+    if elem.isSome():
+      result.add elem.get()
+
+iterator zip*[T1, T2, T3, T4, T5](
+    s1: seq[T1],
+    s2: seq[T2],
+    s3: seq[T3],
+    s4: seq[T4],
+    s5: seq[T5]
+  ): tuple[v1: T1, v2: T2, v3: T3, v4: T4, v5: T5] =
+
+  for idx in 0 ..< min([s1.len, s2.len, s3.len, s4.len, s5.len]):
+    yield (s1[idx], s2[idx], s3[idx], s4[idx], s5[idx])
+
+iterator rmpairs*[T](s: var seq[T]): (int, var T) =
+  ## Iterate over mutable sequence starting from the right
+  for idx in countdown(s.len - 1, 0):
+    yield (idx, s[idx])
+
+
+func escapeStrLit*(input: string): string =
+  input.multiReplace([
+    ("\"", "\\\""),
+    ("\n", "\\n"),
+    ("\\", "\\\\")
+  ])
+
+#*************************************************************************#
+#****************************  Format policy  ****************************#
+#*************************************************************************#
+#===========================  Type definition  ===========================#
+
+import ./dod_helpers
+
+import std/bitops
+declareIdType(LytStr)
+declareHighMasking(LytStr, highMaskRange = 8)
+
+
+type
```

Comment:
A preamble here might help, this is me making a guess, but I'm just trying to draw a "map" of
relationship between the types.

```suggestion
type
  ## These are types for formatting text. Text formatting doesn't have a perfect solution that works
  ## in all cases and we have to resort to heuristics to choose the most approriate text layout on a
  ## case by case basis.
  ##
  ## Text is initially described via a DSL as a set of blocks (`LytBlock`). When it's time to output the
  ## text a layout (`LytLayout`) determines the exact strings (`LytStr`) to output. Text layout requires
  ## heuristics to choose between various options we use a solver ('LytSolution') which is used to
  ## determine the precise layout choosen for a given block.
```